src/tophu/io.py

```diff
@@ -0,0 +1,587 @@
+
+    @overload
+    def __init__(
+        self,
+        filepath: Union[str, os.PathLike],
+        *,
+        band: Optional[int] = None,
+        driver: Optional[str] = None,
+    ):  # noqa: D418
+        """
+        Construct a new `RasterBand` object.
+
+        Parameters
+        ----------
+        filepath : str or path-like
+            Path of the local or remote dataset.
+        band : int or None, optional
+            The (1-based) band index of the raster band. Must be specified if the
+            dataset contains multiple bands. (default: None)
+        driver : str or or None, optional
+            Raster format driver name. If None, registered drivers will be tried
+            sequentially until a match is found. (default: None)
+        """
+        ...
+
+    @overload
+    def __init__(
+        self,
+        filepath: Union[str, os.PathLike],
+        width: int,
+        height: int,
+        dtype: DTypeLike,
+        *,
+        driver: Optional[str] = None,
+        crs: Optional[Union[str, dict, rasterio.crs.CRS]] = None,
+        transform: Optional[rasterio.transform.Affine] = None,
+    ):  # noqa: D418
+        """
+        Construct a new `RasterBand` object.
+
+        Parameters
+        ----------
+        filepath : str or path-like
+            A remote or local dataset path.
+        width, height : int
+            The numbers of columns and rows of the raster dataset.
+        dtype : data-type
+            Data-type of the raster dataset's elements. Must be convertible to a
+            `numpy.dtype` object and must correspond to a valid GDAL datatype.
+        driver : str or None, optional
+            Raster format driver name. If None, the method will attempt to infer the
+            driver from the file extension. (default: None)
+        crs : str, dict, or CRS; optional
+            The coordinate reference system. (default: None)
+        transform : Affine instance, optional
+            Affine transformation mapping the pixel space to geographic space.
+            (default: None)
+        """
+        ...
+
+    def __init__(
+        self,
+        filepath,
+        width=None,
+        height=None,
+        dtype=None,
+        *,
+        band=None,
+        driver=None,
+        crs=None,
+        transform=None,
+    ):  # noqa: D107
+        filepath = Path(filepath)
+
+        # If any of `width`, `height`, or `dtype` are provided, all three must be
+        # provided. If any were not provided, the dataset must already exist. Otherwise,
+        # create the dataset if it did not exist.
+        if (width is None) and (height is None) and (dtype is None):
+            mode = "r"
+            count = None
+        elif (width is not None) and (height is not None) and (dtype is not None):
+            mode = "w+"
+            count = 1
+        else:
+            errmsg = (
+                "the supplied arguments don't match any valid overload of RasterBand"
```

Comment:
```suggestion
                "width, height, and dtype must all be supplied to write a RasterBand"
```

suggesting this because i got confused on my first try to write a file: 
```
In [8]: tophu.io.RasterBand("asdf.tif", width=300, height=200)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[8], line 1
----> 1 tophu.io.RasterBand("asdf.tif", width=300, height=200)

File ~/repos/tophu/src/tophu/io.py:518, in RasterBand.__init__(self, filepath, width, height, dtype, band, driver, crs, transform)
    514 else:
    515     errmsg = (
    516         "the supplied arguments don't match any valid overload of RasterBand"
    517     )
--> 518     raise TypeError(errmsg)
    520 # Create the dataset if it didn't exist.
    521 with rasterio.open(
    522     filepath,
    523     mode,
   (...)
    533     # If a band index was supplied, check that it's within the range of valid
    534     # band indices.

TypeError: the supplied arguments don't match any valid overload of RasterBand
```
but I didn't see what was wrong, and I don't see a way to list the overloads

Reply:
Yeah, I guess I got a little lazy when writing these error messages. Thanks for the heads up that this is a pain point in the UX.

I don't think there's a nice way to inspect function overloads [in Python<3.11](https://docs.python.org/3/library/typing.html#typing.get_overloads). And docstring-parsing tools like pydoc[^pydoc], sphinx[^sphinx], and mkdocstrings[^mkdocstrings] don't yet support `@overload` at the time of writing this as far as I'm aware.

I think what I'll do for now is just hard-code the valid overload signatures in the error message. That seems like the most user-friendly solution, even if it's not ideal for maintainability.

[^pydoc]: [python/cpython#89263](https://github.com/python/cpython/issues/89263)
[^sphinx]: [sphinx-doc/sphinx#7787](https://github.com/sphinx-doc/sphinx/issues/7787)
[^mkdocstrings]: [mkdocstrings/mkdocstrings#308](https://github.com/mkdocstrings/mkdocstrings/issues/308)